tests/app/views/test_questionnaire.py

```diff
@@ -146,3 +146,256 @@ def test_update_questionnaire_store_with_answer_data(self):
+                value=''
+            ), Answer(
+                group_id='who-lives-here',
+                group_instance=0,
+                block_id='household-composition',
+                answer_id='middle-names',
+                answer_instance=1,
+                value=''
+            ), Answer(
+                group_id='who-lives-here',
+                group_instance=0,
+                block_id='household-composition',
+                answer_id='last-name',
+                answer_instance=1,
+                value='Last name only'
+            ),  Answer(
+                group_id='who-lives-here',
+                group_instance=0,
+                block_id='household-composition',
+                answer_id='first-name',
+                answer_instance=2,
+                value='First name only'
+            ), Answer(
+                group_id='who-lives-here',
+                group_instance=0,
+                block_id='household-composition',
+                answer_id='middle-names',
+                answer_instance=2,
+                value=''
+            ), Answer(
+                group_id='who-lives-here',
+                group_instance=0,
+                block_id='household-composition',
+                answer_id='last-name',
+                answer_instance=2,
+                value=''
+            )
+        ]
+
+        answers = []
+        answers.extend(answered)
+        answers.extend(partially_answered)
+
+        for answer in answers:
+            self.question_store.answer_store.add_or_update(answer)
+
+        _remove_empty_household_members_from_answer_store(self.question_store.answer_store, 'who-lives-here')
+
+        for answer in answered:
+            self.assertTrue(self.question_store.answer_store.exists(answer))
+
+        for answer in partially_answered:
+            self.assertTrue(self.question_store.answer_store.exists(answer))
+
+    def test_remove_empty_household_members_middle_name_only_not_stored(self):
+        g.schema_json = load_schema_file("census_household.json")
+
+        unanswered = [
+            Answer(
+                group_id='who-lives-here',
+                group_instance=0,
+                block_id='household-composition',
+                answer_id='first-name',
+                answer_instance=0,
+                value=''
+            ), Answer(
+                group_id='who-lives-here',
+                group_instance=0,
+                block_id='household-composition',
+                answer_id='middle-names',
+                answer_instance=0,
+                value='should not be saved'
+            ), Answer(
+                group_id='who-lives-here',
+                group_instance=0,
+                block_id='household-composition',
+                answer_id='last-name',
+                answer_instance=0,
+                value=''
+            )
+        ]
+
+        answers = []
+        answers.extend(unanswered)
```

Comment:
Could remove the copy here and simply iterate over unanswered in each of the following loops?